### ISSUE TITLE
Build failure with iOS target. Plugin does not seem to locate installed libimobiledevice

### ISSUE THREAD

**Issue by piosystems** [gluonhq/client-maven-plugin#196]
When I run mvn clean client:build with target set to ios, the app is successfully compiled and linked but it never goes beyond. I have posted the error snapshot below

![image](https://user-images.githubusercontent.com/3983248/84954079-a679ec00-b0ec-11ea-9d89-09aa1d4c83ea.png)

Even though I have reinstalled again and again libimobiledevice, I still get the same error as if libimobiledevice has not been installed. My Xcode is version 11 running on Mojave 10.14.6.

Am I missing something?

**Comment by jperedadnr** [comment]
Set `<verbose>true</verbose>` in the plugin's configuration, and run again `mvn client:run`. 

Probably you have a wrong version of libimobiledevice, and uninstalling it might fix it. Else post the console output (as text).

**Comment by piosystems** [comment]
I updated my Xcode to 11.3.1, reinstalled both libusbmuxd and libimobiledevice. Now, not even libusbmuxd is found. See the console output below:

[Thu Jun 18 13:01:02 WAT 2020][INFO] ==================== RUN TASK ====================
[Thu Jun 18 13:01:02 WAT 2020][FINE] PB Command for check brew: which brew
[Thu Jun 18 13:01:02 WAT 2020][FINE] Start process check brew...
[Thu Jun 18 13:01:02 WAT 2020][FINE] [SUB] /usr/local/bin/brew
[Thu Jun 18 13:01:02 WAT 2020][FINE] Result for check brew: 0
[Thu Jun 18 13:01:02 WAT 2020][FINE] Brew found at /usr/local/bin/brew
[Thu Jun 18 13:01:02 WAT 2020][FINE] PB Command for libssl: /bin/sh -c find $(brew --cellar) -name libssl.dylib
[Thu Jun 18 13:01:02 WAT 2020][FINE] Start process libssl...
[Thu Jun 18 13:01:04 WAT 2020][FINE] [SUB] /usr/local/Cellar/openssl@1.1/1.1.1g/lib/libssl.dylib
[Thu Jun 18 13:01:04 WAT 2020][FINE] [SUB] /usr/local/Cellar/openssl/1.0.2p/lib/libssl.dylib
[Thu Jun 18 13:01:04 WAT 2020][FINE] Result for libssl: 0
[Thu Jun 18 13:01:04 WAT 2020][FINE] lib libssl found at /usr/local/Cellar/openssl@1.1/1.1.1g/lib/libssl.dylib
[Thu Jun 18 13:01:04 WAT 2020][FINE] libssl.dylib found in: /usr/local/Cellar/openssl@1.1/1.1.1g/lib/libssl.dylib
[Thu Jun 18 13:01:04 WAT 2020][FINE] lib libssl found at /usr/local/Cellar/openssl/1.0.2p/lib/libssl.dylib
[Thu Jun 18 13:01:04 WAT 2020][FINE] libssl.dylib found in: /usr/local/Cellar/openssl/1.0.2p/lib/libssl.dylib
[Thu Jun 18 13:01:04 WAT 2020][FINE] PB Command for readlink libssl: readlink /usr/local/Cellar/openssl@1.1/1.1.1g/lib/libssl.dylib
[Thu Jun 18 13:01:04 WAT 2020][FINE] Start process readlink libssl...
[Thu Jun 18 13:01:04 WAT 2020][FINE] [SUB] libssl.1.1.dylib
[Thu Jun 18 13:01:04 WAT 2020][FINE] Result for readlink libssl: 0
[Thu Jun 18 13:01:04 WAT 2020][FINE] libssl.dylib link of: libssl.1.1.dylib
[Thu Jun 18 13:01:04 WAT 2020][FINE] PB Command for readlink libssl: readlink /usr/local/Cellar/openssl/1.0.2p/lib/libssl.dylib
[Thu Jun 18 13:01:04 WAT 2020][FINE] Start process readlink libssl...
[Thu Jun 18 13:01:04 WAT 2020][FINE] [SUB] libssl.1.0.0.dylib
[Thu Jun 18 13:01:04 WAT 2020][FINE] Result for readlink libssl: 0
[Thu Jun 18 13:01:04 WAT 2020][FINE] libssl.dylib link of: libssl.1.0.0.dylib
[Thu Jun 18 13:01:04 WAT 2020][FINE] PB Command for libcrypto: /bin/sh -c find $(brew --cellar) -name libcrypto.dylib
[Thu Jun 18 13:01:04 WAT 2020][FINE] Start process libcrypto...
[Thu Jun 18 13:01:05 WAT 2020][FINE] [SUB] /usr/local/Cellar/openssl@1.1/1.1.1g/lib/libcrypto.dylib
[Thu Jun 18 13:01:05 WAT 2020][FINE] [SUB] /usr/local/Cellar/openssl/1.0.2p/lib/libcrypto.dylib
[Thu Jun 18 13:01:05 WAT 2020][FINE] Result for libcrypto: 0
[Thu Jun 18 13:01:05 WAT 2020][FINE] lib libcrypto found at /usr/local/Cellar/openssl@1.1/1.1.1g/lib/libcrypto.dylib
[Thu Jun 18 13:01:05 WAT 2020][FINE] libcrypto.dylib found in: /usr/local/Cellar/openssl@1.1/1.1.1g/lib/libcrypto.dylib
[Thu Jun 18 13:01:05 WAT 2020][FINE] lib libcrypto found at /usr/local/Cellar/openssl/1.0.2p/lib/libcrypto.dylib
[Thu Jun 18 13:01:05 WAT 2020][FINE] libcrypto.dylib found in: /usr/local/Cellar/openssl/1.0.2p/lib/libcrypto.dylib
[Thu Jun 18 13:01:05 WAT 2020][FINE] PB Command for readlink libcrypto: readlink /usr/local/Cellar/openssl@1.1/1.1.1g/lib/libcrypto.dylib
[Thu Jun 18 13:01:05 WAT 2020][FINE] Start process readlink libcrypto...
[Thu Jun 18 13:01:05 WAT 2020][FINE] [SUB] libcrypto.1.1.dylib
[Thu Jun 18 13:01:05 WAT 2020][FINE] Result for readlink libcrypto: 0
[Thu Jun 18 13:01:05 WAT 2020][FINE] libcrypto.dylib link of: libcrypto.1.1.dylib
[Thu Jun 18 13:01:05 WAT 2020][FINE] PB Command for readlink libcrypto: readlink /usr/local/Cellar/openssl/1.0.2p/lib/libcrypto.dylib
[Thu Jun 18 13:01:05 WAT 2020][FINE] Start process readlink libcrypto...
[Thu Jun 18 13:01:05 WAT 2020][FINE] [SUB] libcrypto.1.0.0.dylib
[Thu Jun 18 13:01:05 WAT 2020][FINE] Result for readlink libcrypto: 0
[Thu Jun 18 13:01:05 WAT 2020][FINE] libcrypto.dylib link of: libcrypto.1.0.0.dylib
[Thu Jun 18 13:01:05 WAT 2020][FINE] PB Command for libusbmuxd: /bin/sh -c find $(brew --cellar) -name libusbmuxd.dylib
[Thu Jun 18 13:01:05 WAT 2020][FINE] Start process libusbmuxd...
[Thu Jun 18 13:01:05 WAT 2020][FINE] Result for libusbmuxd: 0
[Thu Jun 18 13:01:05 WAT 2020][FINE] lib libusbmuxd found at 
[Thu Jun 18 13:01:05 WAT 2020][SEVERE] Error finding libusbmuxd.dylib
Check the log files under /Users/piusonobhayedo/Documents/ISM108_2019set/classcodes/hellofx/target/client/arm64-ios/gvm/log
And please check https://docs.gluonhq.com/client/ for more information.
java.lang.RuntimeException: Open a terminal and run the following command to install libusbmuxd.dylib: 

brew install --HEAD libusbmuxd
        at com.gluonhq.substrate.util.ios.Deploy.lambda$checkDependencyPaths$7(Deploy.java:234)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
        at com.gluonhq.substrate.util.ios.Deploy.checkDependencyPaths(Deploy.java:240)
        at com.gluonhq.substrate.util.ios.Deploy.checkPrerequisites(Deploy.java:80)
        at com.gluonhq.substrate.util.ios.Deploy.<init>(Deploy.java:58)
        at com.gluonhq.substrate.target.IosTargetConfiguration.runUntilEnd(IosTargetConfiguration.java:188)
        at com.gluonhq.substrate.SubstrateDispatcher.nativeRun(SubstrateDispatcher.java:484)
        at com.gluonhq.NativeRunMojo.execute(NativeRunMojo.java:45)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:210)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:957)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:289)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:193)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  5.458 s
[INFO] Finished at: 2020-06-18T13:01:05+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.gluonhq:client-maven-plugin:0.1.27:run (default-cli) on project hellofx: Error: Open a terminal and run the following command to install libusbmuxd.dylib: 
[ERROR] 
[ERROR] brew install --HEAD libusbmuxd
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException

**Comment by jperedadnr** [comment]
Did you try `brew install --HEAD libusbmuxd` from a terminal?

**Comment by piosystems** [comment]
Yes.

$ brew install --HEAD libusbmuxd
Warning: libusbmuxd HEAD-c7d7d1a is already installed and up-to-date
To reinstall HEAD, run `brew reinstall libusbmuxd`

**Comment by jperedadnr** [comment]
If you run from terminal: `find $(brew --cellar) -name libusbmuxd.dylib`, what do you get?
In my case I see:
```
/usr/local/Cellar/libusbmuxd/HEAD-e97cebe/lib/libusbmuxd.dylib
```

**Comment by piosystems** [comment]
`find $(brew --cellar) -name libusbmuxd.dylib` returns nothing. But I have seen that the libusbmuxd.dylib exists but with version number as part of name, as can be seen below:

`ls -l /usr/local/Cellar/libusbmuxd/HEAD-c7d7d1a/lib`
 returns:
```
-r--r--r--  1 myusername staff  40864 Jun 18 11:15 libusbmuxd-2.0.6.dylib
-r--r--r--  1 myusername  staff  46304 Jun 18 11:15 libusbmuxd-2.0.a
lrwxr-xr-x  1 myusername  staff     22 Jun 18 11:15 libusbmuxd-2.0.dylib -> libusbmuxd-2.0.6.dylib
```

Similarly, libimobiledevice also exists but with version number as part of name. Thus:

`ls -l /usr/local/Cellar/libimobiledevice/HEAD-4791a82/lib` 
returns
```
-r--r--r--  1 myusername  admin  134084 Jun 18 11:16 libimobiledevice-1.0.6.dylib
-r--r--r--  1 myusername  admin  219520 Jun 18 11:16 libimobiledevice-1.0.a
lrwxr-xr-x  1 myusername  admin      28 Jun 18 11:16 libimobiledevice-1.0.dylib -> libimobiledevice-1.0.6.dylib

```
Thus, only when a * is appended to the `name` passed to `find` will it get it. i.e.

`find $(brew --cellar) -name libusbmuxd*.dylib`
returns
```
/usr/local/Cellar/libusbmuxd/HEAD-c7d7d1a/lib/libusbmuxd-2.0.6.dylib
/usr/local/Cellar/libusbmuxd/HEAD-c7d7d1a/lib/libusbmuxd-2.0.dylib
```
and 

`find $(brew --cellar) -name libimobiledevice*.dylib`
returns
```
/usr/local/Cellar/libimobiledevice/HEAD-4791a82/lib/libimobiledevice-1.0.dylib
/usr/local/Cellar/libimobiledevice/HEAD-4791a82/lib/libimobiledevice-1.0.6.dylib
```

I also wonder why I seem to have multiple versions.

**Comment by jperedadnr** [comment]
Interesting, your libs are being installed with a version number (`libimobiledevice-1.0`, `libusbmuxd-2.0`), so indeed, it looks like you already have other versions. 
Can you do a search for them, and uninstall all of them?

**Comment by jperedadnr** [comment]
After reinstalling, you still get version numbers?
My "head" install is older than yours.
While this get solved, notice that you can still deploy your app via Xcode -> Window -> Devices

**Comment by piosystems** [comment]
I get the impression that `brew install` behaviour has changed. Perhaps the build script which seeks to find the symbolic links needs to change to match.

Both `brew info libusbmuxd` and `brew info libimobiledevice` commands suggest that the latest versions are installed on my system. Yet, the script can't find them.

Meanwhile, I will explore the Xcode -> Window -> Devices

**Comment by piosystems** [comment]
"While this get solved, notice that you can still deploy your app via Xcode -> Window -> Devices"
Can you throw more light on how to go about this?

**Comment by jperedadnr** [comment]
See https://docs.gluonhq.com/gluonvm/#_deploying_to_iphone_ipad_fails_but_the_app_was_created (it was created for an old plugin, but still applies).

**Comment by hlolli** [comment]
Just to add some pressure on this issue, I have the same problem

I tried this nuke trick from https://github.com/flutter/flutter/issues/22595 and also reinstalled homebrew to latest version. Still I keep getting this when trying to run the demo project HelloFx.

```
[Sun Jun 28 23:17:52 CEST 2020][SEVERE] Error finding libusbmuxd.dylib
Check the log files under /Users/hlolli/Documents/client-samples/Maven/HelloFX/target/client/arm64-ios/gvm/log
And please check https://docs.gluonhq.com/client/ for more information.
java.lang.RuntimeException: Open a terminal and run the following command to install libusbmuxd.dylib:

brew install --HEAD libusbmuxd
```

The dylib seems it should be easily discoverable

```
find $(brew --cellar) -name 'libusbmuxd*.dylib'
/usr/local/Cellar/libusbmuxd/HEAD-c7d7d1a/lib/libusbmuxd-2.0.6.dylib
/usr/local/Cellar/libusbmuxd/HEAD-c7d7d1a/lib/libusbmuxd-2.0.dylib
```

**Comment by hlolli** [comment]
just add the wildcard here :P https://github.com/gluonhq/substrate/blob/646a7cfbd5371dbb4e537cfd5a1008fadf1462a9/src/main/java/com/gluonhq/substrate/util/ios/Deploy.java#L224

but this still makes for a shitty reproducibility, I'm a linux user but I hope there's a better way to ask homebrew for installed library than this.

**Comment by hlolli** [comment]
Workaround: 

```
ln -s /usr/local/Cellar/libusbmuxd/HEAD-c7d7d1a/lib/libusbmuxd-2.0.dylib /usr/local/Cellar/libusbmuxd/HEAD-c7d7d1a/lib/libusbmuxd.dylib
```

the `HEAD-c7d7d1a` may differ

**Comment by jperedadnr** [comment]
@hlolli While trying to find out the reason for the "2.0" version, I was following the installation instructions (https://github.com/libimobiledevice/usbmuxd) and ran `brew upgrade libplist`, which upgraded `libusbmuxd.dylib` to `libusbmuxd-2.0.dylib` and `libimobiledevice.dylib` to `libimobiledevice-1.0.dylib` in the process, so it seems they have recently changed their naming conventions... 
We'll provide a fix.

**Comment by jperedadnr** [comment]
It should be fixed now, after https://github.com/gluonhq/substrate/pull/693.

@hlolli if you want to give it a try, you can use the client plugin with 0.1.29-SNAPSHOT version, and this repository:

```
<pluginRepositories>
        <pluginRepository>
            <id>sonatype-snapshots</id>
            <url>https://oss.sonatype.org/content/repositories/snapshots/</url>
        </pluginRepository>
</pluginRepositories>
```

**Comment by piosystems** [comment]
> It should be fixed now, after [gluonhq/substrate#693](https://github.com/gluonhq/substrate/pull/693).
> 
> @hlolli if you want to give it a try, you can use the client plugin with 0.1.29-SNAPSHOT version, and this repository:
> 
> ```
> <pluginRepositories>
>         <pluginRepository>
>             <id>sonatype-snapshots</id>
>             <url>https://oss.sonatype.org/content/repositories/snapshots/</url>
>         </pluginRepository>
> </pluginRepositories>
> ```

Build is now successful; libimobiledevice is now located by the build script

**Comment by piosystems** [comment]
Is there support yet for iOS Simulator in place of physical device?
